web/src/saved_snippets_ui.ts

```diff
@@ -0,0 +1,110 @@
+        $submit_button.prop("disabled", false);
+    }
+}
+
+function saved_snippet_modal_post_render(): void {
+    $("#add-new-saved-snippet-modal").on("input", "input,textarea", update_submit_button_state);
+}
+
+export function rerender_dropdown_widget(): void {
+    const options = saved_snippets.get_options_for_dropdown_widget();
+    saved_snippet_dropdown_widget.list_widget?.replace_list_data(options);
+}
+
+function delete_saved_snippet(saved_snippet_id: string | undefined): void {
+    assert(saved_snippet_id !== undefined);
+    void channel.del({
+        url: "/json/saved_snippet/" + saved_snippet_id,
+    });
+}
+
+function item_click_callback(
+    event: JQuery.ClickEvent,
+    dropdown: tippy.Instance,
+    widget: dropdown_widget.DropdownWidget,
+): void {
+    event.preventDefault();
+    event.stopPropagation();
+
+    if (
+        $(event.target).closest(".saved_snippets-dropdown-list-container .dropdown-list-delete")
+            .length
+    ) {
+        delete_saved_snippet($(event.currentTarget).attr("data-unique-id"));
+        return;
+    }
+
+    dropdown.hide();
+    const current_value = widget.current_value;
+    if (current_value === saved_snippets.ADD_SAVED_SNIPPET_OPTION_ID) {
+        dialog_widget.launch({
+            html_heading: $t_html({defaultMessage: "Add a new saved snippet"}),
+            html_body: render_add_saved_snippet_modal(),
+            html_submit_button: $t_html({defaultMessage: "Save"}),
+            id: "add-new-saved-snippet-modal",
+            form_id: "add-new-saved-snippet-form",
+            update_submit_disabled_state_on_change: true,
+            on_click: submit_create_saved_snippet_form,
+            post_render: saved_snippet_modal_post_render,
+        });
+    } else {
+        const content = saved_snippets
+            .get_saved_snippets()
+            .find((saved_snippet) => saved_snippet.id === current_value)?.content;
+        assert(content !== undefined);
+        const $textarea = $<HTMLTextAreaElement>("textarea#compose-textarea");
+        compose_ui.insert_syntax_and_focus(content, $textarea);
+    }
+}
+
+export const initialize = (params: StateData["saved_snippets"]): void => {
+    saved_snippets.set_saved_snippets(params.saved_snippets);
+
+    saved_snippet_dropdown_widget = new dropdown_widget.DropdownWidget({
+        widget_name: "saved_snippets",
+        get_options: saved_snippets.get_options_for_dropdown_widget,
+        item_click_callback,
+        $events_container: $("body"),
+        unique_id_type: dropdown_widget.DataTypes.NUMBER,
+        focus_target_on_hidden: false,
+        prefer_top_start_placement: true,
+        tippy_props: {
+            // Using -100 as x offset makes saved snippet icon be in the center
+            // of the dropdown widget and 5 as y offset is what we use in compose
+            // recipient dropdown widget.
+            offset: [-100, 5],
+        },
+    });
+    saved_snippet_dropdown_widget.setup();
+};
```

Comment:
I don't think you need to call this function from `ui_init.ts`; you only need this dropdown widget if someone clicks the relevant element in the compose UI, and doing it this way wastes time during web app startup.

The `send_later_popover` code should be the caller for this function, which probably should have a name more like `show`, with care taken to clean up the object when the popover is hidden.

Reply:
There is a error that I am encountering when initializing dropdown widget in `compose_send_menu_popover` -- `saved_snippet_dropdown_widget.list_widget` is getting `undefined` for some reason. I will try to resolve it but is it necessary to avoid initializing at startup? The `compose_recipient_dropdown` is also initialized in `ui_init.ts`.

Reply:
I think we can try to do this in a smaller follow-up PR. It's not essential, just a small initial loading time optimization.